### PR TITLE
[Merged by Bors] - chore: disable relaxedAutoImplicit

### PR DESCRIPTION
**Pull request by kim-em**
We disable the "relaxed" auto-implicit feature, so only single character identifiers become eligible as auto-implicits. See discussion on [zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Auto-implicits) and [2](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.60autoImplicit.20true.60.20is.20evil).

---

- [x] depends on: #5328

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
I've done fancy things with git, and split this into four commits:
* changes to revert after lean4#2277
* actual mathematical errors
* long term changes (but that should still be valid today)
* the actual option change

**Comment by ocfnash** [comment]
Thank you very, very much for doing this and also for going the extra mile with the [PR to Lean](https://github.com/leanprover/lean4/pull/2277) as well as the valuable semantic separation of changes over the four different commits.

For me, the [mathematical errors](https://github.com/leanprover-community/mathlib4/pull/5277/commits/8fab5fcb9e2f38613f3dc38844462ec3ec31c85e) is compelling evidence that we want the change in auto-implicit behaviour proposed here.

Since https://github.com/leanprover/lean4/pull/2277 has already been merged, I suppose we should:
 * bump Mathlib4 to use `leanprover/lean4:nightly-2023-06-??` once it is available [here](https://github.com/leanprover/lean4-nightly/releases)
 * merge the bumped master in here
 * revert [changes that can be reverted after lean4#2277](https://github.com/leanprover-community/mathlib4/pull/5277/commits/9acf7d4b54472c9d5dc476d04c5db7fabeb9751f) (even though personally I'd rather no auto-implicits at all)
 * merge this to master

Or do you want to just merge this to master immediately and bump+revert later?

File: Mathlib/Topology/MetricSpace/Gluing.lean [hunk 1]
```diff
@@ -482,7 +482,7 @@ def GlueSpace (hΦ : Isometry Φ) (hΨ : Isometry Ψ) : Type _ :=
 #align metric.glue_space Metric.GlueSpace
 
 -- porting note: TODO: w/o `@`, tries to generate some `[MetricSpace _]` before finding `X` `Y`
-instance : MetricSpace (@GlueSpace X Y Z _ _ _ _ _ _ hΦ hΨ) :=
+instance (hΦ : Isometry Φ) (hΨ : Isometry Ψ) : MetricSpace (@GlueSpace X Y Z _ _ _ _ _ _ hΦ hΨ) :=
   inferInstanceAs <| MetricSpace <|
```
Comment (kbuzzard) [inline Mathlib/Topology/MetricSpace/Gluing.lean:486]: Is the `@` still needed?

**Review by kbuzzard — APPROVED** [review]
Thanks!

**Comment by semorrison** [comment]
@ocfnash let's just wait until tomorrow, and adapt this to the new Lean4 nightly.

File: Mathlib/Topology/MetricSpace/Gluing.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Topology/MetricSpace/Gluing.lean:486]: Yes, there's no change to this definition.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#5328
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Topology/MetricSpace/Gluing.lean [hunk 2]
```diff
@@ -613,6 +613,10 @@ def InductiveLimit (I : ∀ n, Isometry (f n)) : Type _ :=
   @UniformSpace.SeparationQuotient _ (inductivePremetric I).toUniformSpace
 #align metric.inductive_limit Metric.InductiveLimit
 
+-- We turn off `relaxedAutoImplicit` in mathlib4,
+-- so uppercase single variable implicits need to be declared.
+variable {I}
+
```
Comment (ocfnash) [inline Mathlib/Topology/MetricSpace/Gluing.lean:619]:
```suggestion
```
This can go now, right?

**Review by ocfnash — APPROVED** [review]
Let's get this merged!

bors d+

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**